marilib/marilib.py

```diff
@@ -1,90 +1,125 @@
-            # print(f"Unknown event: {event_type} -- {data}")
-            print("?", end="", flush=True)
+        with self.lock:
+            if len(data) < 1:
+                return
+            self.last_received_serial_data = datetime.now()
+            event_type = data[0]
+
+            if event_type == EdgeEvent.NODE_JOINED:
+                self.gateway.add_node(int.from_bytes(data[1:9], "little"))
+                self.cb_application(EdgeEvent.NODE_JOINED, None)
+            elif event_type == EdgeEvent.NODE_LEFT:
+                if n := self.gateway.remove_node(int.from_bytes(data[1:9], "little")):
+                    self.cb_application(EdgeEvent.NODE_LEFT, n)
+            elif event_type == EdgeEvent.NODE_KEEP_ALIVE:
+                self.gateway.update_node_liveness(int.from_bytes(data[1:9], "little"))
+            elif event_type == EdgeEvent.GATEWAY_INFO:
+                try:
+                    self.gateway.set_info(GatewayInfo().from_bytes(data[1:]))
+                except (ValueError, ProtocolPayloadParserException):
+                    pass
+            elif event_type == EdgeEvent.NODE_DATA:
+                try:
+                    frame = Frame().from_bytes(data[1:])
+                    self.gateway.update_node_liveness(frame.header.source)
+                    payload = frame.payload
+
+                    if payload.startswith(LATENCY_PACKET_MAGIC):
+                        if self.latency_tester:
+                            self.latency_tester.handle_response(frame)
+
+                    if not self._is_test_packet(payload):
+                        self.gateway.register_received_frame(frame)
+
+                    self.cb_application(EdgeEvent.NODE_DATA, frame)
+
+                except (ValueError, ProtocolPayloadParserException):
+                    pass
 
     def send_frame(self, dst: int, payload: bytes):
         assert self.serial_interface is not None
-        mari_frame = Frame(Header(destination=dst), payload=payload)
-        uart_frame_type = b"\x01"
-        uart_frame = uart_frame_type + mari_frame.to_bytes()
-        self.serial_interface.send_data(uart_frame)
-        if node := self.gateway.get_node(dst):
-            node.register_sent_frame(mari_frame)
-        elif dst == MARI_BROADCAST_ADDRESS:
-            for node in self.gateway.nodes:
-                node.register_sent_frame(mari_frame)
-        self.gateway.register_sent_frame(mari_frame)
+
+        with self.lock:
+            if not self._is_test_packet(payload):
+                frame = Frame(Header(destination=dst), payload=payload)
+                if dst == MARI_BROADCAST_ADDRESS:
+                    for n in self.gateway.nodes:
+                        n.register_sent_frame(frame)
+                elif n := self.gateway.get_node(dst):
+                    n.register_sent_frame(frame)
+                self.gateway.register_sent_frame(frame)
+
+        frame_to_send = Frame(Header(destination=dst), payload=payload)
```

Comment:
I think you could create the `frame` only once in the beginning.
Also, can you add a quick one-sentence comment explaining why there is a different handling for test packets and normal packets.
Then it's ready to merge!